### DISCUSSION TITLE
Add Separate Environment Variable for Public S3 Endpoint in Self-Hosted Setup

### DISCUSSION THREAD

**Discussion by eloko7** [langfuse/langfuse#4716] [Ideas]
### Describe the feature or potential improvement

Currently, when running Langfuse v3 in a self-hosted Docker setup, the UI uses `LANGFUSE_S3_MEDIA_UPLOAD_ENDPOINT` to generate URLs for media assets (such as images from multimodal generation). While this endpoint works internally (`http://minio:9000`), it’s not accessible from outside the Docker network, resulting in broken images in the UI.

**Issue:**

- `LANGFUSE_S3_MEDIA_UPLOAD_ENDPOINT` serves a dual purpose: internal write and external read.
- When set to an internal Docker hostname (`http://minio:9000`), media uploads work internally, but the UI-generated URLs are not publicly accessible.
- Changing the endpoint to `http://localhost:9090` fixes external access but may break internal references and presigned URLs used by the server.

**Proposed Solution:**

Introduce a new environment variable (e.g. `LANGFUSE_S3_MEDIA_PUBLIC_ENDPOINT`) that `langfuse-web` uses when generating UI-facing URLs. This would allow:

- `LANGFUSE_S3_MEDIA_UPLOAD_ENDPOINT` to remain set to the internal endpoint (`http://minio:9000`).
- `LANGFUSE_S3_MEDIA_PUBLIC_ENDPOINT` to be set to a host-accessible endpoint (e.g. `http://localhost:9090`).

**Benefits:**

- Clean separation of internal and external endpoints.
- No more broken images in the UI when running self-hosted versions with Docker.

**Conclusion:**

By adding `LANGFUSE_S3_MEDIA_PUBLIC_ENDPOINT`, operators can smoothly handle internal and external media endpoints, ensuring both internal services and the end-user UI have proper access to uploaded media.

**Comment by hassiebp** [comment]
Hi @eloko7 - thanks a lot for raising this idea! Have you seen our [docs section on this specific issue already?](https://langfuse.com/self-hosting/docker-compose#configuring-minio-for-media-uploads)

> To enable multimodal tracing, presigned URLs allow SDK clients and browsers outside the Docker network to directly upload and download media assets. Therefore, the LANGFUSE_S3_MEDIA_UPLOAD_ENDPOINT must resolve to the Docker host’s address.

> Development Environment: When running docker compose locally, set LANGFUSE_S3_MEDIA_UPLOAD_ENDPOINT to http://localhost:9090 to ensure presigned URLs correctly loop back to the local instance.

> Production Environment: In a production environment, configure LANGFUSE_S3_MEDIA_UPLOAD_ENDPOINT with a publicly accessible hostname or IP address that is reachable by your SDK clients and browsers.

By the abovementioned approach, an additional environment variable would not be necessary. Please let me know whether you feel I'm missing something!
